### PR TITLE
Add streaming support

### PR DESCRIPTION
**Pull request by geoffrey-brier**

### CODE REVIEW

File: src/Widop/HttpAdapter/BuzzHttpAdapter.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ public function __construct(Browser $browser = null, $maxRedirects = 5)
     /**
      * {@inheritdoc}
      */
-    public function getContent($url, array $headers = array())
+    public function getContent($url, array $headers = array(), $streamCallback = null)
```
Comment (egeloen) [inline src/Widop/HttpAdapter/BuzzHttpAdapter.php]: IMO, `$streamCallback` is not a good naming choice especially when there is already a `StreamHttpAdapter` in the library. What about `$persistentCallback`?

**Comment by egeloen** [comment]
@geoffrey-brier What does the callback receive as arguments? Are there always the same accross all http adapters?

File: src/Widop/HttpAdapter/BuzzHttpAdapter.php (hunk 1 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/BuzzHttpAdapter.php]: That makes sense ^^

File: src/Widop/HttpAdapter/BuzzHttpAdapter.php [hunk 2]
```diff
@@ -83,9 +89,23 @@ public function getName()
...
+     * @throws \Widop\HttpAdapter\HttpAdapterException If client is not supported for persistent callback.
      */
-    private function configure()
+    private function configure($persistentCallback)
     {
-        $this->browser->getClient()->setMaxRedirects($this->getMaxRedirects());
+        $client = $this->browser->getClient();
+
+        if ($persistentCallback !== null) {
+            if (!$client instanceof Curl) {
```
Comment (egeloen) [inline src/Widop/HttpAdapter/BuzzHttpAdapter.php]: At the first read, I'm :-1: here (especially for the case of Buzz). If we start making this kind of check, we should do the same for files as Buzz can't make a post multipart send with the default adapter (file get contents). Additionally, why do you restrict it to Curl and not AbstractCurl?

IMO, it should be documented that multipart post & persistent connection are only supported with Buzz if the AbstractCurl adapter is used but nothing more. What's your opinion?

File: src/Widop/HttpAdapter/CurlHttpAdapter.php [hunk 3]
```diff
@@ -46,21 +51,27 @@ public function postContent($url, array $headers = array(), array $content = arr
...
      *
-     * @param string   $url      A valid URL.
-     * @param array    $headers  Http headers.
-     * @param array    $content  Http content (in case of POST method).
-     * @param callable $callback A callable function.
+     * @param string   $url            A valid URL.
+     * @param array    $headers        Http headers.
+     * @param array    $content        Http content (in case of POST method).
+     * @param callable $callback       A callable function.
+     * @param callable $persistentCallback A callable function if you want to stream the content.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/CurlHttpAdapter.php]: The persistent callback + CS alignment

File: src/Widop/HttpAdapter/CurlHttpAdapter.php [hunk 4]
```diff
@@ -46,21 +51,27 @@ public function postContent($url, array $headers = array(), array $content = arr
...
      *
-     * @return string The response content.
+     * @return string|null The response content or null in case of streaming.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/CurlHttpAdapter.php]: in case of persistent connection

File: src/Widop/HttpAdapter/GuzzleHttpAdapter.php [hunk 5]
```diff
@@ -87,10 +92,17 @@ public function getName()
     /**
      * Configures the guzzle request.
      *
-     * @param \Guzzle\Http\Message\RequestInterface $request The request.
+     * @param \Guzzle\Http\Message\RequestInterface $request        The request.
+     * @param callable                              $persistentCallback The stream callable.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/GuzzleHttpAdapter.php]: The persistent callback + CS alignment

File: src/Widop/HttpAdapter/HttpAdapterInterface.php [hunk 6]
```diff
@@ -26,28 +26,36 @@
     /**
      * Gets the content fetched from the given URL.
      *
-     * @param string $url     The URL to request.
-     * @param array  $headers HTTP headers (optionnal).
+     * @param string   $url            The URL to request.
+     * @param array    $headers        HTTP headers (optionnal).
+     * @param callable $persistentCallback The callable.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: The persistent callback + CS alignment

File: src/Widop/HttpAdapter/HttpAdapterInterface.php [hunk 7]
```diff
@@ -26,28 +26,36 @@
...
      *
-     * @param string $url     The URL to request.
-     * @param array  $headers HTTP headers (optional).
-     * @param array  $content The POST content (optional).
-     * @param array  $files   The POST files (optional).
+     * @param string   $url            The URL to request.
+     * @param array    $headers        HTTP headers (optional).
+     * @param array    $content        The POST content (optional).
+     * @param array    $files          The POST files (optional).
+     * @param callable $persistentCallback The callable.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: The persistent callback + CS alignment

File: src/Widop/HttpAdapter/HttpAdapterInterface.php [hunk 8]
```diff
@@ -26,28 +26,36 @@
...
      *
-     * @param string $url     The URL to request.
-     * @param array  $headers HTTP headers (optionnal).
+     * @param string   $url            The URL to request.
+     * @param array    $headers        HTTP headers (optionnal).
+     * @param callable $persistentCallback The callable.
      *
      * @throws \Widop\HttpAdapter\HttpAdapterException If an error occured.
      *
      * @return string The fetched content.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: string|null

File: src/Widop/HttpAdapter/HttpAdapterInterface.php [hunk 9]
```diff
@@ -26,28 +26,36 @@
...
-     * @param array  $files   The POST files (optional).
+     * @param string   $url            The URL to request.
+     * @param array    $headers        HTTP headers (optional).
+     * @param array    $content        The POST content (optional).
+     * @param array    $files          The POST files (optional).
+     * @param callable $persistentCallback The callable.
      *
      * @throws \Widop\HttpAdapter\HttpAdapterException If an error occured.
      *
      * @return string The fetched content.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: string|null

File: src/Widop/HttpAdapter/StreamHttpAdapter.php [hunk 10]
```diff
@@ -24,40 +24,52 @@ class StreamHttpAdapter extends AbstractHttpAdapter
...
     }
 
     /**
      * Calls an URL given a context.
      *
-     * @param string   $url     An url.
-     * @param resource $context A resource (created from stream_context_create).
+     * @param string   $url            An url.
+     * @param resource $context        A resource (created from stream_context_create).
+     * @param callable $persistentCallback A stream callback.
```
Comment (egeloen) [inline src/Widop/HttpAdapter/StreamHttpAdapter.php]: The persistent callback + CS alignment

**Comment by egeloen** [comment]
The zend http adapter needs some love too & the build is broken

File: src/Widop/HttpAdapter/BuzzHttpAdapter.php (hunk 2 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/BuzzHttpAdapter.php]: Ok, as we've discussed we'll document it.

File: src/Widop/HttpAdapter/CurlHttpAdapter.php (hunk 4 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/CurlHttpAdapter.php]: Fixed

File: src/Widop/HttpAdapter/CurlHttpAdapter.php (hunk 3 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/CurlHttpAdapter.php]: Fixed

File: src/Widop/HttpAdapter/GuzzleHttpAdapter.php (hunk 5 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/GuzzleHttpAdapter.php]: Fixed

File: src/Widop/HttpAdapter/HttpAdapterInterface.php (hunk 6 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: Fixed

File: src/Widop/HttpAdapter/HttpAdapterInterface.php (hunk 8 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: Fixed

File: src/Widop/HttpAdapter/HttpAdapterInterface.php (hunk 7 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: Fixed

File: src/Widop/HttpAdapter/HttpAdapterInterface.php (hunk 9 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/HttpAdapterInterface.php]: Fixed

File: src/Widop/HttpAdapter/StreamHttpAdapter.php (hunk 10 above)
Comment (geoffrey-brier) [inline src/Widop/HttpAdapter/StreamHttpAdapter.php]: Fixed

**Comment by egeloen** [comment]
@geoffrey-brier I have rationalize persitent callback to only receive datas & it does not need to return the exact input size now (handle on top of it).

For the Zend adapter, I have take a look to the documentation & only found the option `persistent` in this table http://framework.zend.com/manual/2.2/en/modules/zend.http.client.adapters.html#zend-http-client-adapter-socket-configuration-table but I don't know how we can receive datas incrementally...

Last part, the PR needs some tests.

**Comment by geoffrey-brier** [comment]
Ok sounds good to me.
We need to investigate the Zend case and if it turns out that it accepts a persistent callback then we'll need to update the docs and the class too. I'll try and take a look tonight.

**Comment by egeloen** [comment]
@geoffrey-brier I have started to rebase this PR but there is still one question: Why does all adapters return an http response when we use the streaming support?

**Pull request closed without merging**